KH-Amir-TN/greenlight
# GIT-3100: Reconsidered the password complexity solution design

## Description

As of the latest release of Greenlight (GL) `v2.11.x` the password complexity has increased.
The solution design used a flag (`secure_password`) to mark accounts with "secure/insecure" passwords so that the application can decide the proper action to take with each account.
Any user local account with insecure password (`secure_password == false`) won't login and gets redirected directly to `reset_password` page to update it matching the new requirements.
This raised some issues as described in #3078,#3079,#3081, #3091,#3100.
This addresses #3100 and #3091 while PR #3082 targets the others.

## Solution:
The solution is straightforward, any local account will have its password validated against the complexity requirements upon login:
1. If the password is secure then the user login.
2. If the password is insecure (doesn't match the `PASSWORD_PATTERN`) the user won't login and they will get redirected to reset password page.
This will ensure that old accounts with already "secure" passwords doesn't get prompted to change their password.

The password minimal complexity required remains the same:
1. A password has to have at least **One digit**.
2. A password has to have at least **One non-word characters (anything but `a-z, A-Z, 0-9`)**.
3. A password has to have at least **One Upper case letter**.
4. A password has to have at least **One lower case letter**.
5. A password has to be at least **8 characters in length**.

## Testing Steps

1. Ran `bigbluebutton/greenlight:v2.10.0.3` on empty database.
2. Created three accounts with insecure password.
3. Opened three browser sessions for each account.
4. Ran this PR docker image on the existing database.
5. Attempted to change the password on one account with insecure password:
   5.1 Confirmed the complexity validation for the new password.
   5.2 Changed the password to a secure one.
   5.3 Signed out and resigned in on same account and confirming that the user doesn't has to reset their password.
6. Signed out all of the other sessions and resigned in to confirm that:
   6.1 The account with insecure password can't login and gets redirected to reset password page:
       6.1.1 All validation for password complexity works.
       6.1.2 After resetting the password and signing in the user logs in without having to reset their password.
       6.1.3 Signing out and resigning in without having to reset the password.
   6.2 The user with secure password logs in without having to reset their password.

Also, Tested also for `user` and `admin` create rake tasks.
## Screenshots (if appropriate):
